Mathlib/Analysis/Normed/Group/SeparationQuotient.lean

```diff
@@ -0,0 +1,161 @@
+/-
+Copyright (c) 2024 Yoh Tanimoto. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Yoh Tanimoto
+-/
+import Mathlib.Analysis.Normed.Group.Hom
+
+/-!
+# Lifts of maps to separation quotients of seminormed groups
+
+For any `SeminormedAddCommGroup M`, a `NormedAddCommGroup` instance has been defined in
+`Mathlib.Analysis.Normed.Group.Uniform`.
+
+## Main definitions
+
+We use `M` and `N` to denote seminormed groups.
+All the following definitions are in the `SeparationQuotient` namespace. Hence we can access
+`SeparationQuotient.normedMk` as `normedMk`.
+
+* `normedMk` : the normed group hom from `M` to `SeparationQuotient M`.
+
+* `liftNormedAddGroupHom` : implements the universal property of `SeparationQuotient M`. Here
+    `(f : NormedAddGroupHom M N)`, `(hf : ∀ x : M, ‖x‖ = 0 → f x = 0)` and
+    `liftNormedAddGroupHom f hf : NormedAddGroupHom (SeparationQuotient M) N`.
+
+## Main results
+
+* `norm_normedMk` : the operator norm of the projection is `1` if the subspace is not dense.
+
+* `norm_liftNormedAddGroupHom_le` : `‖liftNormedAddGroupHom f hf‖ ≤ ‖f‖`.
+
+## Implementation details
+
+For any `SeminormedAddCommGroup M`, we define a norm on `SeparationQuotient M` by
+`‖x‖ = ‖mk x‖` using the lift.
+
+-/
+
+
+noncomputable section
+
+open SeparationQuotient NNReal
+
+variable {M N : Type*} [SeminormedAddCommGroup M] [SeminormedAddCommGroup N]
+
+namespace SeparationQuotient
+
+/-- For `(m : M)`, `mk m = 0` if and only if `‖m‖ = 0`. -/
+theorem mk_eq_zero_iff (m : M) : mk m = 0 ↔ ‖m‖ = 0 := by
+  rw [← norm_mk, norm_eq_zero]
+
+open NormedAddGroupHom
+
+/-- The morphism from a seminormed group to the quotient by the inseparable setoid. -/
+noncomputable def normedMk : NormedAddGroupHom M (SeparationQuotient M) where
+  __ := mkAddMonoidHom
+  bound' := ⟨1, fun m => by simp only [ZeroHom.toFun_eq_coe, AddMonoidHom.toZeroHom_coe,
+    mkAddMonoidHom_apply, norm_mk, one_mul, le_refl]⟩
+
+/-- `normedMk` agrees with `SeparationQuotient.mk`. -/
+@[simp]
+theorem normedMk.apply (m : M) : normedMk m = mk m := rfl
+
+/-- The operator norm of the projection is at most `1`. -/
+theorem norm_normedMk_le : ‖normedMk (M := M)‖ ≤ 1 :=
+  NormedAddGroupHom.opNorm_le_bound _ zero_le_one fun m => by simp only [normedMk.apply, norm_mk,
+    one_mul, le_refl]
+
+lemma eq_of_inseparable (f : NormedAddGroupHom M N) (hf : ∀ x, ‖x‖ = 0 → f x = 0) :
+    ∀ x y, Inseparable x y → f x = f y :=
+  fun x y h ↦ eq_of_sub_eq_zero <| by
+    rw [← map_sub]
+    rw [Metric.inseparable_iff, dist_eq_norm] at h
+    exact hf (x - y) h
+
+/-- The lift of a group hom to the separation quotient as a group hom. -/
+@[simps]
+noncomputable def liftNormedAddGroupHom (f : NormedAddGroupHom M N)
+    (hf : ∀ x, ‖x‖ = 0 → f x = 0) : NormedAddGroupHom (SeparationQuotient M) N :=
+  { SeparationQuotient.liftContinuousAddMonoidHom ⟨f.toAddMonoidHom, f.continuous⟩
```

Comment:
Do we not have  `NormedAddGroupHom.toContinuousAddMonoidHom`?

Reply:
apparently not... are you suggesting to add an abbrev?

Reply:
No, we should have `ContinuousMonoidHomClass.toContinuousMonoidHom` (with `@[to_additive]`) as a coercion, and then you can just appeal to this coercion instead. Note however that the assumptions should be 
```lean
(F : Type*) [FunLike F A B] [ContinuousMapClass F A B] [MonoidHomClass F A B] (f : F
```
since `ContinuousMonoidHomClass` is deprecated.

Reply:
I did it, plus `ContinuousMapClass (NormedAddGroupHom V₁ V₂) V₁ V₂`. I added `toContinuousMonoidHom` as `CoeTC`, because with `Coe` I got the error that synthesizing order could not be determined. Is this a right thing?